src/aarch64/ffi.c

```diff
@@ -728,34 +737,95 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+    ~__builtin_cheri_representable_alignment_mask(stack_bytes),
+    _Alignof(stack));
+  frame_padding = max(
+    ~__builtin_cheri_representable_alignment_mask(sizeof(*frame)),
+    _Alignof(*frame));
+  rvalue_padding = max(~__builtin_cheri_representable_alignment_mask(rsize),
+    _Alignof(rvalue));
+#endif
+  context = alloca (sizeof(struct call_context) +
+#ifdef __CHERI_PURE_CAPABILITY__
+		    stack_padding +
+#endif
+		    stack_bytes +
+#ifdef __CHERI_PURE_CAPABILITY__
+		    frame_padding +
+#endif
+		    sizeof(struct call_frame) +
+#ifdef __CHERI_PURE_CAPABILITY__
+		    rvalue_padding +
+#endif
+		    rsize);
   _Static_assert(sizeof(struct call_context) == CALL_CONTEXT_SIZE, "");
   _Static_assert(sizeof(struct call_frame) == CALL_FRAME_SIZE, "");
-  stack = context + 1;
 #ifdef __CHERI_PURE_CAPABILITY__
   /*
-   * context is bounded to the alloca size, we want to pass a pointer that
-   * points just after the alloca with the curretn stack bounds.
+   * Align the stack so that the c9 capability to on-stack function arguments
+   * can be bounded exactly to the number of bytes occupied by them.
    */
-  stack = __builtin_cheri_address_set(__builtin_cheri_stack_get(), (ptraddr_t)stack);
-#endif
+  cursor = (uintptr_t)context + sizeof(*context);
+  stack = (void*)__align_up(cursor,
+    ~__builtin_cheri_representable_alignment_mask(stack_bytes) + 1);
+  stack = (void*)__align_up(stack, _Alignof(stack));
+  stack = __builtin_cheri_bounds_set_exact(stack,
+    __builtin_cheri_round_representable_length(stack_bytes));
+
+  cursor += stack_padding + stack_bytes;
+  frame = (void*)__align_up(cursor,
+    ~__builtin_cheri_representable_alignment_mask(sizeof(*frame)) + 1);
+  frame = (void*)__align_up(frame, _Alignof(*frame));
+  frame = __builtin_cheri_bounds_set_exact(frame, sizeof(*frame));
+#else
+  stack = context + 1;
   frame = (void*)((uintptr_t)stack + (uintptr_t)stack_bytes);
+#endif
   /* Initialize the frame with 0xaa to detect errors. */
   memset(context, 0xaa, sizeof(*context));
   memset(frame, 0xaa, sizeof(*frame));
   if (rsize) {
-    /* frame currently has full stack bounds, so we can use it for derivation. */
-    rvalue = (void*)((uintptr_t)frame + sizeof(struct call_frame));
 #ifdef __CHERI_PURE_CAPABILITY__
-    rvalue = __builtin_cheri_bounds_set(rvalue, rsize);
+    cursor += frame_padding + sizeof(*frame);
+    rvalue = __align_up(cursor,
+      ~__builtin_cheri_representable_alignment_mask(rsize) + 1);
+    rvalue = __align_up(rvalue, _Alignof(rvalue));
+    rvalue = __builtin_cheri_bounds_set_exact(rvalue, rsize);
+#else
+    rvalue = (void*)((uintptr_t)frame + sizeof(struct call_frame));
 #endif
   } else {
     rvalue = orig_rvalue;
   }
 #ifdef __CHERI_PURE_CAPABILITY__
-  frame = __builtin_cheri_bounds_set(frame, sizeof(*frame));
+  /*
+   * All capabilities are derived from context at this point. Set the bounds
+   * now.
+   */
+  context = __builtin_cheri_bounds_set_exact(context, sizeof(*context));
```

Comment:
You can do this immediately after `cursor = (uintptr_t)context + sizeof(*context);`?

Reply:
I wanted to separate this from handling particular sub-objects, but I guess the current version using macros is much cleaner.